### PR TITLE
[RFC] Add a way to color command-line

### PR DESCRIPTION
**Pull request by ZyX-I**
Based on #5119 as I do not want merge conflicts.

The idea is that we can color Neovim command-line. Ultimate goal is

1. Refactor VimL parsing so it can produce information about colors (for `:`/command prompts).
2. Refactor VimL expression parsing so it can produce informaiton about colors (for `=`/expr prompts).
3. Add a way for the user to specify how his input() prompts should be colored.
4. Add a way for the user to specify how his custom commands should be colored.

This is logical to split that into the following PRs:

1. (This) add basic prompt coloring functionality. Adds a few callbacks, including one which may be used to make pygments color Ex commands and expressions for us. Also add a way for the user to color input() prompts. If some strange problems will not occur this part is going to be easy.
2. Remove callback which colors expressions, move expressions parsing to `src/nvim/eval` and split it from execution, make it produce coloring information. Obsoletes or reuses expression parser from #243. Reuses tests from #243. Resulting code is going to be rather unoptimal: expressions are going to be reparsed each time.
3. (Many PRs) port commands parsing from #243; most likely one-by-one since #243 is not good for highlighting and is rather outdated. Execute commands with C, no translation. Again reparse things each time up until all commands are ported.
4. Actually build AST for functions and execute it without reparsing. Though this has nothing to do with command-line coloring.
5. Only now provide `:command -color=funcname` for coloring user commands.

---

Proposed API for input():

`input(prompt :: String[, text :: String[, completion :: String]])`: no additions.

`input(prompt :: String, options :: Dictionary)`, `options` allow keys text, completion and highlight. No third argument allowed.

### CODE REVIEW

**Comment by ZyX-I** [comment]
PR commits  start with “ex_getln: Clean up draw_cmdline a bit”.

**Comment by ZyX-I** [comment]
Example code which will highlight using pygments:

```Python
# color_cmdline.py
def color_cmdline_setup():
    import vim
    from pygments import lex, format
    from pygments.lexers import get_lexer_by_name
    from pygments.formatter import Formatter
    from pygments.token import Token

    try:
        from __builtin__ import unicode
    except ImportError:
        unicode = str

    COLORS = {
        Token:                    'Normal',

        Token.Comment:            'Comment',
        Token.Comment.Preproc:    'PreProc',
        Token.Keyword:            'Keyword',
        Token.Keyword.Type:       'Type',
        Token.Operator:           'Operator',
        Token.Name:               'Identifier',
        Token.Name.Builtin:       'Statement',
        Token.Name.Function:      'Function',
        Token.Name.Constant:      'Number',
        Token.String:             'String',
        Token.Number:             'Number',
        Token.Punctuation:        'Operator',

        Token.Error:              'Error',
    }

    class NvimFormatter(Formatter):
        def __init__(self, **kwargs):
            super(NvimFormatter, self).__init__(**kwargs)
            self.offset = 0
            self.colorscheme = kwargs.get('colorscheme', None) or COLORS

        def format_unencoded(self, tokensource, outfile):
            for ttype, value in tokensource:
                ttype_str = str(ttype).replace(' ', '_')
                value = value.rstrip('\n')
                color = self.colorscheme.get(ttype)
                while color is None:
                    ttype = ttype[:-1]
                    color = self.colorscheme.get(ttype)
                if isinstance(value, unicode):
                    vlen = len(value.encode('utf-8'))
                else:
                    vlen = len(value)
                if vlen:
                    outfile.write('{} {} {} {}\n'.format(
                        self.offset, self.offset + vlen, color, ttype_str))
                self.offset += vlen

    def setup():
        global color_cmdline_highlight
        lexer = get_lexer_by_name('vim')

        def color_cmdline_highlight():
            cmdline = vim.eval('a:cmdline')
            add_tt = False
            if bool(int(vim.eval('get(a:000, 0)'))):
                add_tt = True
            tokens = lex(cmdline, lexer)
            fmt = format(tokens, NvimFormatter(bg=vim.options['background']))
            return [
                ((lambda s, e, g, tt: (
                    (int(s), int(e), g)
                    + ((tt,) if add_tt else ())
                ))(*line.split()))
                for line in fmt.split('\n')[:-1]
            ]

    setup()


color_cmdline_setup()
```

```Vim
pyfile <sfile>:h/color_cmdline.py
function PygmentsFormat(cmdline, ...)
  return pyeval('color_cmdline_highlight()')
endfunction
let g:Nvim_color_cmdline = 'PygmentsFormat'
```

**Comment by bfredl** [comment]
it might be worthwhile to think of how this will interact with #6162 later. I guess GUIs can recieve a list of pairs `[text, attrs]` with the attrs that apply to that chunk of text, alternatively just the text and the list of`[start, stop, attrs]`.

File: test/functional/ui/cmdline_highlight_spec.lua [hunk 1]
```diff
@@ -55,47 +119,227 @@ describe('Command-line coloring', function()
...
+      :echo "                                 |
+      {ERR:E5406: Chunk 0 end 7 splits multibyte ch}|
+      {ERR:aracter}                                 |
+      :echo "«                                |
+      {ERR:E5406: Chunk 0 end 7 splits multibyte ch}|
+      {ERR:aracter}                                 |
+      :echo "«^                                |
+    ]])
+  end)
+  it('does the right thing when errorring', function()
```
Comment (brcolow) [inline test/functional/ui/cmdline_highlight_spec.lua]: typo: errorring -> erroring

**Comment by ZyX-I** [comment]
Actually it appears that `input({opts})` is better then requiring prompt. What’s the point in requiring prompt if the official documentation states that you may have no prompt by supplying it an empty string, so basically prompt is not required?

**Comment by bfredl** [comment]
I would guess the purpose is to nudge plugin authors to consider adding a prompt (instead of just calling `input()` with no args without even checking the docs that a prompt is possible), but with a dict they will need to check the docs anyway so then it makes sense to make it optional.

Also, having user commands and `input()` executing a callback on every input change is going to be useful for other purposes than coloring the cmdline, for instance live feedback in a window like `inccommand`. So I wonder if the callback argument really should be called `highlight` or `color`, maybe instead `callback` to acknowledge that it has other potential uses?

**Comment by ZyX-I** [comment]
@bfredl One of the points of the coloring is “better have no colors then make user problems”. People using that for other purposes should be prepared that Neovim may reject calling their callback if it causes troubles: shows errors, takes too long to finish, alters some state, explicitly disabled by the user, etc. In the PR I am only going to disable it for errors, arabic shaping (not sure how to implement and how to test) and probably for some option set, but that may be extended depending on the problems I spot. If you know an existing facility to automatically interrupt after some time passed I would implement “too long to finish” as well.

**Comment by bfredl** [comment]
> alters some state, explicitly disabled by the user, etc. In the PR I am only going to disable it for errors, arabic shaping (not sure how to implement and how to test) and probably for some option set, 

This would be fine for other use cases as long as one still can `rpcnotify()` (not `rpcrequest()`) and `start_timer(0, ...)` (currently the canonical way of "do something when the main loop is ready")

> If you know an existing facility to automatically interrupt

I don't know if `got_int` somehow can be abused for this. Checking `got_int` is quite common but calling `os_breakcheck` less so (which would be the natural place to check the time limit)

File: src/nvim/ex_getln.h [hunk 2]
```diff
@@ -52,6 +52,8 @@ typedef struct hist_entry {
   list_T *additional_elements;  ///< Additional entries from ShaDa file.
 } histentry_T;
 
+extern Callback getln_input_callback;
```
Comment (bfredl) [inline src/nvim/ex_getln.h]: why not make it part of `save_ccline` ?

**Comment by ZyX-I** [comment]
Should be mostly ready. The only issue left I know about is making `<C-c>` interrupt test less flaky. Maybe just adding a sleep will be enough.

**Comment by ZyX-I** [comment]
QB has the usual kbtree problem and that’s the only failure. AppVeyor got

```
[  FAILED  ] C:/projects/neovim/test/functional\core\job_spec.lua @ 267: jobs can pass user data to the callback
C:/projects/neovim/test/functional\core\job_spec.lua:271: Expected objects to be the same.
Passed in:
(table) {
  [1] = 'notification'
  [2] = 'stdout'
 *[3] = {
    [1] = {
      [l] = { ... more }
      [n] = 5
      [s] = 'str' }
   *[2] = {
      [1] = 'foo' } } }
Expected:
(table) {
  [1] = 'notification'
  [2] = 'stdout'
 *[3] = {
    [1] = {
      [l] = { ... more }
      [n] = 5
      [s] = 'str' }
   *[2] = {
      [1] = 'foo'
     *[2] = '' } } }
stack traceback:
	C:/projects/neovim/test/functional\core\job_spec.lua:271: in function <C:/projects/neovim/test/functional\core\job_spec.lua:267>
```

yet again, otherwise OK. Travis is still building, but so far everything succeeded and only coverage was left.

**Comment by ZyX-I** [comment]
Travis succeeded.

File: runtime/doc/eval.txt [hunk 3]
```diff
@@ -4723,6 +4724,23 @@ input({opts})
...
+		where
+			hl_start_byte is the first highlighted byte,
+			hl_end_byte is the last highlighted byte (+ 1!),
+			hl_group is |:hl| group used for highlighting.
+					      *E5403* *E5404* *E5405* *E5406*
+		Both hl_start_byte and hl_end_byte + 1 must point to the start 
+		of the multibyte character (highlighting must not break 
+		multibyte characters), hl_end_byte + 1 may be equal to the 
+		input length.  Start column must be in range [0, len(input)), 
+		end column must be in range (hl_start_byte, len(input)],
```
Comment (bfredl) [inline runtime/doc/eval.txt]: I don't think `(,]` notation is used elsewhere. Maybe "end byte must be greater than start byte in the same item". Also, always use "byte" and not mix "byte" with "column"?

File: src/nvim/api/private/helpers.c [hunk 4]
```diff
@@ -37,6 +37,64 @@ typedef struct {
...
+  FUNC_ATTR_NONNULL_ALL FUNC_ATTR_WARN_UNUSED_RESULT
+{
+  const bool ret = !try_end(err);
+  assert(trylevel == 0);
+  assert(!need_rethrow);
+  assert(!got_int);
+  assert(!did_throw);
+  assert(!did_emsg);
+  assert(msg_list == &tstate->private_msg_list);
+  assert(*msg_list == NULL);
```
Comment (bfredl) [inline src/nvim/api/private/helpers.c]: Is it intentional, if compiled without assertions, they will be silently ignored? (including leaking memory?)

File: test/functional/eval/input_spec.lua [hunk 5]
```diff
@@ -196,6 +228,19 @@ describe('input()', function()
     eq('Vim(call):E118: Too many arguments for function: input',
        exc_exec('call input("prompt> ", "default", "file", "extra")'))
   end)
+  it('supports highlighting', function()
+    feed([[:call input({"highlight": "RainBowParens"})<CR>]])
+    wait()
+    feed('(())')
+    wait()
```
Comment (bfredl) [inline test/functional/eval/input_spec.lua]: wait shouldn't be used prior to `expect()` (I'm also suspicious of the one before, `screen:sleep()` is better if we really need to simulate "the user stopped typing for some while", but I wouldn't expect that here, `input()` should be sync in its operations)

File: src/nvim/ex_getln.c [hunk 6]
```diff
@@ -2374,8 +2641,23 @@ static void draw_cmdline(int start, int len)
     }
 
     msg_outtrans_len(arshape_buf, newlen);
-  } else
-    msg_outtrans_len(ccline.cmdbuff + start, len);
+  } else {
+draw_cmdline_no_arabicshape:
```
Comment (bfredl) [inline src/nvim/ex_getln.c]: should we document that coloring is disabled with arabicshape? Or is it a limitiation we expect to remove soon?

File: runtime/doc/eval.txt (hunk 3 above)
Comment (ZyX-I) [inline runtime/doc/eval.txt]: This is ISO 31-11 standard mathematical notation for sets. Writing this in words would be rather lengthy and I pretty much sure I would not like the result.

File: src/nvim/api/private/helpers.c (hunk 4 above)
Comment (ZyX-I) [inline src/nvim/api/private/helpers.c]: Where do you see leaking memory? No assertions here are doing anything, but checking const comparisons results.

File: runtime/doc/eval.txt (hunk 3 above)
Comment (bfredl) [inline runtime/doc/eval.txt]: but is it used elsewhere in the documentation? You write a similar constraint in words below and it doesn't seem to lengthy.

File: src/nvim/api/private/helpers.c (hunk 4 above)
Comment (bfredl) [inline src/nvim/api/private/helpers.c]: I mean if the disabled asserts would have failed, we will silently overwrite this state, including pointers that I presume points to allocated memory?

File: runtime/doc/eval.txt (hunk 3 above)
Comment (ZyX-I) [inline runtime/doc/eval.txt]: “Similar constraint” below is mere ≥. And it *is* lengthy: compare “≥” or “[ to “is greater then or equal to”, just not *too* lengthy and I also do not like writing “previous {varname}” in a math expression, it looks bad and not mathy.

File: src/nvim/api/private/helpers.c (hunk 4 above)
Comment (ZyX-I) [inline src/nvim/api/private/helpers.c]: They are, but if asserts would’ve failed we are getting some kind of “undefined behaviour” and not “leaking memory”: failing assert indicates that code is being based on wrong assumptions. Fixing “memory leak” in this case will most likely just introduce another bug instead, like silently ignored exception.

File: test/functional/eval/input_spec.lua (hunk 5 above)
Comment (ZyX-I) [inline test/functional/eval/input_spec.lua]: Without the one before test fails.

File: src/nvim/api/private/helpers.c (hunk 4 above)
Comment (bfredl) [inline src/nvim/api/private/helpers.c]: That's what meant with "intentional", we can accept undefined exception state and small memory leak until the code elsewhere is fixed.

File: test/functional/eval/input_spec.lua (hunk 5 above)
Comment (bfredl) [inline test/functional/eval/input_spec.lua]: do we know why? does `input()` flush the input buffer somehow?  Real interaction will not have this API call, so it would be good to identify the issue.

Comment (ZyX-I) [inline test/functional/eval/input_spec.lua]: @bfredl I got that when I removed all the `wait()`s:

```
./test/functional/ui/screen.lua:302: Row 4 did not match.
Expected:
  |{EOB:~                        }|
  |{EOB:~                        }|
  |{EOB:~                        }|
  |*                         |
  |{RBP1:(}{RBP2:()}{RBP1:)}^                     |
Actual:
  |{EOB:~                        }|
  |{EOB:~                        }|
  |{EOB:~                        }|
  |*{RBP1:(}{RBP2:()}{RBP1:)}^                     |
  |                         |

To print the expect() call that would assert the current screen state, use
screen:snapshot_util(). In case of non-deterministic failures, use
screen:redraw_debug() to show all intermediate screen states.

stack traceback:
        ./test/functional/ui/screen.lua:302: in function 'wait'
        ./test/functional/ui/screen.lua:216: in function 'expect'
        test/functional/eval/input_spec.lua:233: in function <test/functional/eval/input_spec.lua:230>
```

I do not know why though. And I do not see any problems in TUI.

Comment (ZyX-I) [inline test/functional/eval/input_spec.lua]: After adding a logger which checks how my function is being called I see that it is being called twice for each input. It is rather strange, this should indicate that `draw_cmdline` is called twice.

Comment (ZyX-I) [inline test/functional/eval/input_spec.lua]: May be a good idea to cache previous highlighting result and reuse it if prompt_id and text to highlight matches what was used previously.

Comment (bfredl) [inline test/functional/eval/input_spec.lua]: is it not just that the cmdline does not fit the screen width? If I make the terminal narrow I get the same behavior in TUI.

File: runtime/doc/eval.txt (hunk 3 above)
Comment (bfredl) [inline runtime/doc/eval.txt]: My point is mainly of consistency: there is no need to use obscure math notation to save a few words in otherwise spelled out paragraph text (which is what I comment on, and not an equation)

Comment (ZyX-I) [inline runtime/doc/eval.txt]: It is a standard math notation, why do you say it is obscure? It saves lots of words, not “a few”. And it is already used in documentation: with incorrectly (does not find my own interval) constructed regex there is a match in `:h log()`. Actually, there is more then one occurrence in math functions’ descriptions.

File: test/functional/eval/input_spec.lua (hunk 5 above)
Comment (ZyX-I) [inline test/functional/eval/input_spec.lua]: @bfredl In the test where I observed calling that twice I had 118 columns, no prompt and one to four characters (typed `(())` with a slightly adjusted RainBowParens function). In the input_spec while screen is rather narrow it still fits.

Comment (bfredl) [inline test/functional/eval/input_spec.lua]: I was referring to the extra `wait()`,  which will not be needed if remove the narrow width issue (make the window wider, or use `command`)

I've also noticed that the cmdline redraw logic is not very efficient. We needed some throttling to implement the cursor type switching without overloading the ui protocol serialization with data....

Comment (ZyX-I) [inline test/functional/eval/input_spec.lua]: I guess I understood: what is being too wide is the previous command, not the input. In the other (new) test file I omitted the issue by creating a mapping which calls `input`. I cannot use `command` though, it blocks.

Comment (bfredl) [inline test/functional/eval/input_spec.lua]: you can use `nvim_async('command', ...)`

File: src/nvim/message.c [hunk 7]
```diff
@@ -1628,6 +1628,27 @@ void msg_puts_attr_len(const char *const str, const ptrdiff_t len, int attr)
...
 
+/// Print a formatted message
+///
+/// Message printed is limited by #IOSIZE. Must not be used from inside
+/// msg_puts_attr().
+///
+/// @param[in]  attr  Highlight attributes.
+/// @param[in]  fmt  Format string.
+void msg_printf_attr(const int attr, const char *const fmt, ...)
+  FUNC_ATTR_NONNULL_ALL
```
Comment (oni-link) [inline src/nvim/message.c]: This attribute looks too restrictive. What if we want to print a pointer value `NULL`?

**Comment by ZyX-I** [comment]
I have added caching of the highlight callback calling results. Pay attention to `-- (…)` in new tests: it appears that highlight callback (with command-line redrawing function) may be called *three* times in a row with one and the same cmdline.

**Comment by ZyX-I** [comment]
Travis hanged again.

File: runtime/doc/vim_diff.txt [hunk 8]
```diff
@@ -126,7 +126,6 @@ Commands:
 Functions:
   |dictwatcheradd()| notifies a callback whenever a |Dict| is modified
   |dictwatcherdel()|
-  |execute()| works with |:redir|
```
Comment (justinmk) [inline runtime/doc/vim_diff.txt]: why was this removed?

Comment (ZyX-I) [inline runtime/doc/vim_diff.txt]: Vim has `execute()`, so it can’t be in vim_diff any longer. And in any case implementation has nothing to do with `:redir`.

Comment (justinmk) [inline runtime/doc/vim_diff.txt]: The behavior of `execute(':redir')` differs. That's what this note is about.

Comment (ZyX-I) [inline runtime/doc/vim_diff.txt]: Then what is it doing in “New features” section?

**Comment by ZyX-I** [comment]
Travis passed.

File: src/nvim/api/private/helpers.c [hunk 9]
```diff
@@ -37,6 +37,64 @@ typedef struct {
...
+  current_exception = tstate->current_exception;
+  trylevel = tstate->trylevel;
+  got_int = tstate->got_int;
+  did_throw = tstate->did_throw;
+  need_rethrow = tstate->need_rethrow;
+  did_emsg = tstate->did_emsg;
+  return ret;
+}
+
 /// Start block that may cause vimscript exceptions
```
Comment (bfredl) [inline src/nvim/api/private/helpers.c]: Maybe clarify when this can be used as opposed to `try_enter`. IIUC command line could be colored in a situation where there could be a pending error (perhaps from a job/timer event handler active in cmdline mode?) so `try_enter` must be used, while an API function can only be called in a so far error free state, so `try_start` is safe?

File: test/functional/eval/input_spec.lua [hunk 10]
```diff
@@ -196,6 +228,18 @@ describe('input()', function()
     eq('Vim(call):E118: Too many arguments for function: input',
        exc_exec('call input("prompt> ", "default", "file", "extra")'))
   end)
+  it('supports highlighting', function()
+    feed([[:call input({"highlight": "RainBowParens"})<CR>]])
+    wait()
```
Comment (bfredl) [inline test/functional/eval/input_spec.lua]: this wait should still be removed. In any realistic situation, `input()` will be invoked with one call, not with an unrelated event incidentally right after it.

**Comment by bfredl** [comment]
Otherwise LGTM

File: src/nvim/api/private/helpers.c (hunk 9 above)
Comment (ZyX-I) [inline src/nvim/api/private/helpers.c]: Pending errors I saw so far were either from the command-line itself (there is a test with error in `:while` cycle body) or from things like mappings (there also is a test). If timers/etc may spawn errors like this then they should use `try_enter` as well.

**Comment by ZyX-I** [comment]
(Without the merge clint fails locally.)

**Comment by ZyX-I** [comment]
AppVeyor failure looks unrelated:

```
[  FAILED  ] 1 test, listed below:
[  FAILED  ] C:/projects/neovim/test/functional\core\job_spec.lua @ 69: jobs changes to given `cwd` directory
C:/projects/neovim/test/functional\core\job_spec.lua:78: Expected objects to be the same.
Passed in:
(table) {
  [1] = 'notification'
  [2] = 'stdout'
 *[3] = {
    [1] = 0
   *[2] = {
      [1] = 'C:\projects\neovim\Xtest-tmpdir\nvimdDir0n\0' } } }
Expected:
(table) {
  [1] = 'notification'
  [2] = 'stdout'
 *[3] = {
    [1] = 0
   *[2] = {
      [1] = 'C:\projects\neovim\Xtest-tmpdir\nvimdDir0n\0'
     *[2] = '' } } }
stack traceback:
	C:/projects/neovim/test/functional\core\job_spec.lua:78: in function <C:/projects/neovim/test/functional\core\job_spec.lua:69>
```

Travis hanged in oldtests, test49:

```
!echo 'q^Mq' | $NVIM_PRG -u NONE -N -es -c 'debuggreedy|set viminfo+=nviminfo' -c 'let ExtraVimBegin = 1' -c 'let ExtraVimResult = "/tmp/nvimSdwGqV/9"' -c 'breakadd func 3 G' -S /tmp/nvimSdwGqV/8
No output has been received in the last 10m0s, this potentially indicates a stalled build or something wrong with the build itself.
```

Restarting.

**Comment by justinmk** [comment]
My tentative observation is that the test49 hang seems to persist if travis cache is not cleared.

**Comment by ZyX-I** [comment]
@justinmk This is planned to be merged after dealing with more urgent issues for the next (0.2.1) release?

**Comment by justinmk** [comment]
@ZyX-I If you think it's ready, I think we should include it in 0.2.1

File: runtime/doc/vim_diff.txt [hunk 11]
```diff
@@ -265,6 +273,8 @@ Lua interface (|if_lua.txt|):
 on cancel and completion respectively) via dictionary argument (replaces all 
 other arguments if used).
 
+|input()| and |inputdialog()| now support user-defined cmdline highlighting.
```
Comment (justinmk) [inline runtime/doc/vim_diff.txt]: Why is only highlighting mentioned, in all of these docs? This is a handler that is called on each input character, correct? That's useful for much more than highlighting.

File: runtime/doc/eval.txt [hunk 12]
```diff
@@ -4703,6 +4703,7 @@ input({opts})
 		cancelreturn  ""       Same as {cancelreturn} from
 		                       |inputdialog()|. Also works with 
 		                       input().
+		highlight     nothing  Highlight handler: |Funcref|.
```
Comment (justinmk) [inline runtime/doc/eval.txt]: How often is the handler called? Should mention this somewhere.

Typically we prefix VimL handlers with `on_` (Vim does `_cb` suffix). Maybe this should be called `on_input`.

File: runtime/doc/vim_diff.txt (hunk 11 above)
Comment (ZyX-I) [inline runtime/doc/vim_diff.txt]: @bfredl asked this somewhere above: it is not guaranteed to be called, if real hooks cause any problems then I would rather add code which detects those problems and stops using hook then do anything. Currently this includes stopping calling the hook in case of arabic characters present. And handler is not called on each input character, it is called on *redraw*, minus cases when hook output can be taken from cache. I don’t have any information regarding what caused redraw, so one will have to compute a diff: hook always receives the whole drawn command-line. And whatever other uses you have in mind, note that hook is not supposed to edit input, `:echo` anything or issue any errors.

File: runtime/doc/vim_diff.txt [hunk 13]
```diff
@@ -147,6 +146,15 @@ Highlight groups:
   |hl-TermCursorNC|
   |hl-Whitespace| highlights 'listchars' whitespace
 
+UI:
+		*E5408* *E5409* *g:Nvim_color_expr* *g:Nvim_color_cmdline*
+  UI now supports command-line coloring. Officially only |input()| and 
+  |inputdialog()| may be colored, temporary for testing purposes expressions 
+  (e.g. |i_CTRL-R_=|) and regular command-line (|:|) are colored by callbacks 
+  defined in `g:Nvim_color_expr` and `g:Nvim_color_cmdline` respectively.
```
Comment (justinmk) [inline runtime/doc/vim_diff.txt]: `g:Nvim_` might not be a good convention because `sessionoptions+=globals` will try to store such variables if they are string/num types (doesn't apply here, but as a general convention ...).

Since these are temporary and deprecated, perhaps underscore prefix would make sense.

Comment (ZyX-I) [inline runtime/doc/vim_diff.txt]: @justinmk I can’t have leading underscores, funcref variable names must start with a *capital*. Underscore is not one.

File: test/functional/ui/cmdline_highlight_spec.lua [hunk 14]
```diff
@@ -0,0 +1,848 @@
...
+local meths = helpers.meths
+local funcs = helpers.funcs
+local source = helpers.source
+local dedent = helpers.dedent
+local command = helpers.command
+local curbufmeths = helpers.curbufmeths
+
+local screen
+
+-- Bug in input() handling: {REDRAW} will erase the whole prompt up until
```
Comment (justinmk) [inline test/functional/ui/cmdline_highlight_spec.lua]: should say `:redraw!` instead of `{REDRAW}` in this context.

File: runtime/doc/eval.txt (hunk 12 above)
Comment (ZyX-I) [inline runtime/doc/eval.txt]: `on_input` is not right, it is currently called on redraw. And I do not like the hook or event view of the function, in my view it is more like a syntax definition, in a form of the VimL code: this allows things like caching in case of double highlight I implemented, additionally I don’t know command-line redrawing system enough to answer questions like “how many times and for which states highlight will be called when triggering `a` mapping defined like `cnoremap a bcd<C-\>e"efg"<CR>`” without experiments.

File: src/nvim/ex_getln.c [hunk 15]
```diff
@@ -2285,75 +2343,329 @@ void free_cmdline_buf(void)
...
+    goto color_cmdline_end;
+  }
+  if (colored_ccline->highlight_callback.type != kCallbackNone) {
+    // Currently this should only happen while processing input() prompts.
+    assert(colored_ccline->input_fn);
+    color_cb = colored_ccline->highlight_callback;
+  } else if (colored_ccline->cmdfirstc == ':') {
+    try_enter(&tstate);
+    err_errmsg = N_(
+        "E5408: Unable to get Nvim_color_cmdline callback from g:: %s");
```
Comment (justinmk) [inline src/nvim/ex_getln.c]: `g::` looks awkward. Suggestion:

    "E5408: Unable to get g:Nvim_color_cmdline callback: %s"

File: runtime/doc/vim_diff.txt (hunk 11 above)
Comment (ZyX-I) [inline runtime/doc/vim_diff.txt]: Though no, arabic characters fall under different category: “hard to implement, unlikely to happen and should really be solved on the different layer and not with some hacks in ex_getln.c and screen.c (it contains another incarnation of arabic shaping code)”. What falls into “causes problems” category now are hooks which throw errors.

**Comment by justinmk** [comment]
There's a non-trivial amount of logic involving `getln_interrupted_highlight`, `prompt_id`, related to concerns about performance. Just curious, why would performance be noticed here, typically commands occupy only 1 line, and AFAIK vim does not take such measures for insert-mode highlighting.

**Review by justinmk — APPROVED** [review]

**Comment by ZyX-I** [comment]
Insert-mode highlighting is done in C code. I am running VimL and expect it to be slow in cases like somebody trying to implement pygments in pure VimL. And I do not see too non-trivial amount: for `getln_interrupted_highlight` it is just “if `<C-c>` happened to arrive while highlighting assume that highlighting appeared to be too slow for the user to be disturbed, so no longer execute the highlight (`:try` catches interrupts as exceptions, so nothing is done specially here, they go with other errors), but do not assume that user meant to interrupt the prompt (what `getln_interrupted_highlight` is for)”. `prompt_id` is there not for performance, it disables running highlight in case of errors (in order to not spam user with errors), but makes the disablement not entirely permanent, only for one prompt.

**Comment by ZyX-I** [comment]
Or, the other way around: `getln_interrupted_highlight` is what allows you to interrupt infinite cycle in highlight function, but *still* run what you want. Without it you could not do `let g:Nvim_color_cmdline = function('Halting')` and run any Ex mode command afterwards without hacks (there still is `<C-r>=` and `execute()`), including the one which will fix the problem.

`prompt_id` allows disabling highlighting for the duration of one prompt only, in the above case it is needed for you to not type `<C-c>` after typing each character of `unlet g:Nvim_color_cmdline`. Though I think I can remove `prompt_id` and just keep `prev_prompt_errors` and `last_ccline_colors` in `CmdlineInfo`. Behaviour will slightly change: in the example you would need to type second `<C-c>` if you use another prompt: `:u<C-c>nlet <C-r>="g"<CR>:<C-c>Nvim_color_cmdline<CR>`, and highlight function will not be recalled.

**Comment by ZyX-I** [comment]
CI succeeded, only left running is gcov “flaky” build which will not possibly affect green status.

**Comment by ZyX-I** [comment]
All “flaky” appeared to be green as well, it is uncommon.
